research-software-directory/RSD-as-a-service
# 171 scraper logs

# Store scraper error messages and logs in database

Changes proposed in this pull request:

* Git and package manager scrapers now store user friendly error messages and full stack traces in the database when an exception occurs.
* Git and package manager only scrape entries that were never scraped  or scraped at least one hour ago to prevent recurring error messages when there are only a few entries in the database

How to test:

* `docker compose down --volumes && docker compose build --parallel && docker-compose up --scale data-generation=0`
* Create a software page with a non-existing repo URL and a non-existing package manager URL
* Run the scrapers:
```bash
docker compose exec scrapers java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.git.MainProgrammingLanguages
docker compose exec scrapers java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.git.MainContributors
docker compose exec scrapers java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.git.MainBasicData
docker compose exec scrapers java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.git.MainCommits
docker compose exec scrapers java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.package_manager.MainPackageManager
```
* Check in the database that appropriate error messages and logs were created: `SELECT * FROM repository_url;`, `SELECT * FROM package_manager;` and `SELECT * FROM backend_log;`
* Run the scrapers again, because they ran less than an hour ago, nothing should happen
* Because of row level security, nothing should show at http://localhost/api/v1/backend_log
* Now enter a valid repository URL and package manager URL, run the scrapers again and check that they stored real data


Related to #171, but doesn't close it as the frontend isn't adapted yet

PR Checklist:
